### ISSUE TITLE
Add response status code to transaction metadata

### ISSUE THREAD

**Issue by matsimitsu** [appsignal/appsignal-ruby#183]
Response status codes (200,202 etc.) could help debug performance issues. e.g. short-returning a 404-not found could have a different impact on your app then returning a 200 with a body. While it might be clear from the event timeline, I think it's nice to have as additional metadata. 

---

Blocked by https://github.com/appsignal/appsignal-ruby/issues/329

## Instrumentations

Response status reporting will be supported in https://github.com/appsignal/appsignal-ruby/pull/1090 through the Rack EventHandler. This will only work for Rails first, as it's the first instrumentation to use that. As we port other instrumentations over to use Rack EventHandler, more libraries will be supported.

- [x] pure-Rack - EventHandler
- [x] Rails #1090 
- [x] Sinatra - #1100
- [x] Hanami - #1113
- [x] Grape - needs docs change
- [x] Padrino - https://github.com/appsignal/appsignal-ruby/pull/1142
- [x] ~Webmachine?~ Not doing this

**Comment by tombruijn** [comment]
We can get the status code from the `@app.call(env)` response like in the code below when we would add it to the [rails_instrumentation.rb](https://github.com/appsignal/appsignal-ruby/blob/12a052929f4ddf4ebc5ea0ba0a44376051a3aaf0/lib/appsignal/rack/rails_instrumentation.rb#L28)

```ruby
@app.call(env).tap do |status, _, __|
  transaction.set_metadata("status", status)
end
```

This would work for any Rack application since they are required to return the response triplet: status code, headers, and body (in that order). So we can add it for Rack, Sinatra and Grape too.

Is this what you had in mind, @matsimitsu ?

**Comment by matsimitsu** [comment]
Sounds good! 👍

**Comment by phoet** [comment]
@tombruijn so this is implemented for the elexir client but not in the ruby version?

**Comment by tombruijn** [comment]
@phoet correct. For us to capture all response codes we need to rework how our middleware works. Currently we would only get part of the response codes. When there’s middleware before AppSignal’s we would not track them, giving an incomplete picture.

We want to improve that in issue https://github.com/appsignal/appsignal-ruby/issues/329

**Comment by richardonrails** [comment]
This would be very helpful. I was hoping to see throughput and response time graphs by HTTP response code out-of-the box with AppSignal.

**Comment by tombruijn** [comment]
Requested in this [Intercom conversation](https://app.intercom.com/a/inbox/yzor8gyw/inbox/admin/4356044/conversation/16410700334511#part_id=comment-16410700334511-21909563610).

Working on this in PR #1090

**Comment by tombruijn** [comment]
We report the `response_status` metric for the response status since Ruby gem 3.8.1 for Rails apps. Other frameworks coming soon. You can now create a graph on a dashboard to show this metric.

**Comment by phoet** [comment]
💋

**Comment by phoet** [comment]
Hm, I'm not able to create a dashboard with metrics from `response_status`. I do see the status-code in the performance samples as a tag though 🤔

**Comment by tombruijn** [comment]
You should be able to make a graph like this: 
![image](https://github.com/appsignal/appsignal-ruby/assets/282402/e4c90cf9-9f3a-4722-9409-4e99b0836322)

If the tag is set on samples, the metric is sent as well. These things are [set right after one another](https://github.com/appsignal/appsignal-ruby/blob/7c376b03ea2266ae84946dc93f8111bdf9c83faa/lib/appsignal/rack/event_handler.rb#L64-L71).

**Comment by phoet** [comment]
ah, i was missing the namespace part 👍

**Comment by tombruijn** [comment]
This has been released in several recent Ruby gems, with the last libraries supported in Ruby gem 3.10.
